compiler/rustc_hir_typeck/src/fallback.rs

```diff
@@ -160,6 +167,61 @@ impl<'tcx> FnCtxt<'_, 'tcx> {
         true
     }
 
+    /// Existing code relies on `f32: From<T>` (usually written as `T: Into<f32>`) resolving `T` to
+    /// `f32` when the type of `T` is inferred from an unsuffixed float literal. Using the default
+    /// fallback of `f64`, this would break when adding `impl From<f16> for f32`, as there are now
+    /// two float type which could be `T`, meaning that the fallback of `f64` would be used and
+    /// compilation error would occur as `f32` does not implement `From<f64>`. To avoid breaking
+    /// existing code, we instead fallback `T` to `f32` when there is a trait predicate
+    /// `f32: From<T>`. This means code like the following will continue to compile:
+    ///
+    /// ```rust
+    /// fn foo<T: Into<f32>>(_: T) {}
+    ///
+    /// foo(1.0);
+    /// ```
+    fn calculate_fallback_to_f32(&self, unresolved_variables: &[Ty<'tcx>]) -> UnordSet<FloatVid> {
+        let Some(from_trait) = self.tcx.lang_items().from_trait() else {
+            return UnordSet::new();
+        };
+        let pending_obligations = self.fulfillment_cx.borrow_mut().pending_obligations();
+        debug!("calculate_fallback_to_f32: pending_obligations={:?}", pending_obligations);
+        let roots: UnordSet<ty::FloatVid> = pending_obligations
+            .into_iter()
+            .filter_map(|obligation| {
+                // The predicates we are looking for look like
+                // `TraitPredicate(<f32 as std::convert::From<{float}>>, polarity:Positive)`.
+                // They will have no bound variables.
+                obligation.predicate.kind().no_bound_vars()
+            })
+            .filter_map(|predicate| match predicate {
+                ty::PredicateKind::Clause(ClauseKind::Trait(TraitPredicate {
+                    polarity: PredicatePolarity::Positive,
+                    trait_ref,
+                })) if trait_ref.def_id == from_trait
+                    && self.shallow_resolve(trait_ref.self_ty()).kind()
+                        == &ty::Float(ty::FloatTy::F32) =>
+                {
+                    self.root_float_vid(trait_ref.args.type_at(1))
+                }
+                _ => None,
+            })
+            .collect();
+        debug!("calculate_fallback_to_f32: roots={:?}", roots);
+        if roots.is_empty() {
+            // Most functions have no `f32: From<{float}>` predicates, so short-circuit and return
+            // an empty set when this is the case.
+            return UnordSet::new();
+        }
+        let fallback_to_f32 = unresolved_variables
+            .iter()
+            .flat_map(|ty| ty.float_vid())
+            .filter(|vid| roots.contains(&self.root_float_var(*vid)))
+            .collect();
```

Comment:
Why do the mapping from root to all unified infer vars here instead of looking at the root in `fallback_if_possible`? Please add a comment